data/syncer/userAccountsSyncer.go

```diff
@@ -0,0 +1,164 @@
+	requestHandler     trie.RequestHandler
+	waitTime           time.Duration
+	shardId            uint32
+	cacher             storage.Cacher
+	rootHash           []byte
+}
+
+// ArgsNewUserAccountsSyncer defines the arguments needed for the new account syncer
+type ArgsNewUserAccountsSyncer struct {
+	Hasher             hashing.Hasher
+	Marshalizer        marshal.Marshalizer
+	TrieStorageManager data.StorageManager
+	RequestHandler     trie.RequestHandler
+	WaitTime           time.Duration
+	ShardId            uint32
+	Cacher             storage.Cacher
+}
+
+// NewUserAccountsSyncer creates a user account syncer
+func NewUserAccountsSyncer(args ArgsNewUserAccountsSyncer) (*userAccountsSyncer, error) {
+	if check.IfNil(args.Hasher) {
+		return nil, state.ErrNilHasher
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, state.ErrNilMarshalizer
+	}
+	if check.IfNil(args.TrieStorageManager) {
+		return nil, state.ErrNilStorageManager
+	}
+	if check.IfNil(args.RequestHandler) {
+		return nil, state.ErrNilRequestHandler
+	}
+	if args.WaitTime < time.Second {
+		return nil, state.ErrInvalidWaitTime
+	}
+
+	u := &userAccountsSyncer{
+		hasher:             args.Hasher,
+		marshalizer:        args.Marshalizer,
+		trieSyncers:        make(map[string]data.TrieSyncer),
+		dataTries:          make(map[string]data.Trie),
+		trieStorageManager: args.TrieStorageManager,
+		requestHandler:     args.RequestHandler,
+		waitTime:           args.WaitTime,
+		shardId:            args.ShardId,
+		cacher:             args.Cacher,
+		rootHash:           nil,
+	}
+
+	return u, nil
+}
+
+// SyncAccounts will launch the syncing method to gather all the data needed for userAccounts
+func (u *userAccountsSyncer) SyncAccounts(rootHash []byte) error {
+	u.rootHash = rootHash
+
+	dataTrie, err := trie.NewTrie(u.trieStorageManager, u.marshalizer, u.hasher)
+	if err != nil {
+		return err
+	}
+
+	u.dataTries[string(rootHash)] = dataTrie
+	trieSyncer, err := trie.NewTrieSyncer(u.requestHandler, u.cacher, dataTrie, u.waitTime, u.shardId, factory.AccountTrieNodesTopic)
+	if err != nil {
+		return err
+	}
+	u.trieSyncers[string(rootHash)] = trieSyncer
+
+	err = trieSyncer.StartSyncing(rootHash)
+	if err != nil {
+		return err
+	}
+
+	rootHashes, err := u.findAllAccountRootHashes(dataTrie)
+	if err != nil {
+		return err
+	}
+
+	err = u.syncAccountDataTries(rootHashes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *userAccountsSyncer) syncAccountDataTries(rootHashes [][]byte) error {
+	for _, rootHash := range rootHashes {
+		dataTrie, err := trie.NewTrie(u.trieStorageManager, u.marshalizer, u.hasher)
+		if err != nil {
+			return err
+		}
+
+		u.dataTries[string(rootHash)] = dataTrie
+		trieSyncer, err := trie.NewTrieSyncer(u.requestHandler, u.cacher, dataTrie, u.waitTime, u.shardId, factory.AccountTrieNodesTopic)
+		if err != nil {
+			return err
+		}
+		u.trieSyncers[string(rootHash)] = trieSyncer
+
+		err = trieSyncer.StartSyncing(rootHash)
```

Comment:
Maybe add a task to change the trie syncer behavior: if, for any particular reason, it fails to fetch a trie node, it will return an error instead of retrying.

Reply:
added